python/ray/serve/_private/autoscaling_state.py

```diff
@@ -0,0 +1,355 @@
+        requests.
+        """
+
+        total_requests = 0
+
+        if (
+            RAY_SERVE_COLLECT_AUTOSCALING_METRICS_ON_HANDLE
+            or len(self._running_replicas) == 0
+        ):
+            for handle_metric in self._handle_requests.values():
+                total_requests += handle_metric.queued_requests
+                for id in self._running_replicas:
+                    if id in handle_metric.running_requests:
+                        total_requests += handle_metric.running_requests[id]
+        else:
+            for id in self._running_replicas:
+                if id in self._replica_average_ongoing_requests:
+                    total_requests += self._replica_average_ongoing_requests[id][1]
+
+        return total_requests
+
+
+class AutoscalingStateManager:
+    """Manages all things autoscaling related.
+
+    Keeps track of request metrics for each deployment and decides on
+    the target number of replicas to autoscale to based on those metrics.
+    """
+
+    def __init__(self):
+        self._autoscaling_states: Dict[DeploymentID, AutoscalingState] = {}
+
+    def get_target_num_replicas(
+        self, deployment_id: DeploymentID, curr_target_num_replicas: int
+    ) -> int:
+        return self._autoscaling_states[deployment_id].get_decision_num_replicas(
+            curr_target_num_replicas=curr_target_num_replicas,
+        )
+
+    def register_deployment(
+        self,
+        deployment_id: DeploymentID,
+        info: DeploymentInfo,
+        curr_target_num_replicas: Optional[int] = None,
+    ):
+        """If deployment is an autoscaling deployment, register its info.
+        Otherwise, remove it from tracking.
+        """
+
+        if info.deployment_config.autoscaling_config:
+            if deployment_id not in self._autoscaling_states:
+                self._autoscaling_states[deployment_id] = AutoscalingState(
+                    deployment_id
+                )
+            return self._autoscaling_states[deployment_id].register(
+                info, curr_target_num_replicas
+            )
+        else:
+            if deployment_id in self._autoscaling_states:
+                del self._autoscaling_states[deployment_id]
+
+    def deregister_deployment(self, deployment_id: DeploymentID):
+        """Remove deployment from tracking."""
+
+        if deployment_id in self._autoscaling_states:
+            del self._autoscaling_states[deployment_id]
+
+    def update_running_replica_ids(
+        self, deployment_id: DeploymentID, running_replicas: List[ReplicaID]
+    ):
+        if deployment_id in self._autoscaling_states:
+            self._autoscaling_states[deployment_id].update_running_replica_ids(
+                running_replicas
+            )
+
+    def get_total_num_requests(self, deployment_id: DeploymentID) -> float:
+        return self._autoscaling_states[deployment_id].get_total_num_requests()
```

Comment:
Do we need to guard this with an `if deployment_id in self._autoscaling_states` check like the other methods (such as `update_running_replica_ids`)?

Reply:
Same question for the rest of the methods (e.g. `is_within_bounds`, `apply_bounds`, etc).

Reply:
These methods are all only called in deployment state manager's `autoscale()` which is only called if the deployment id is tracked in the autoscaling state managejr.